doc/Architecture/Sexp.org

```diff
@@ -0,0 +1,399 @@
+1. basic
+   - input
+     #+begin_src haskell
+       let foo =
+         mod Bar =
+           let bar = 3
+           type zar = Boo
+         end in
+         Bar.bar
+     #+end_src
+   - output
+     #+begin_src lisp
+       (:defun foo ()
+         (:let-mod Bar ()
+            ((:defun bar () 3)
+             (type zar () (Boo)))
+           Bar.bar))
+     #+end_src
+*** Tuples
+1. basic
+   - input
+     #+begin_src haskell
+       let foo = (1,2,3)
+     #+end_src
+   - output
+     #+begin_src lisp
+       (:defun foo () (:tuple 1 2 3))
+     #+end_src
+*** Lists
+1. basic
+   - input
+     #+begin_src haskell
+       let foo = [1,2,3,4]
+     #+end_src
+   - output
+     #+begin_src lisp
+       (:defun foo () (:list 1 2 3 4))
+     #+end_src
+*** Records
+1. basic
+   - input
+     #+begin_src haskell
+       let foo = {a, b = 2}
+     #+end_src
+   - output
+     #+begin_src lisp
+       (:defun foo () (:record (a) (b 2)))
+     #+end_src
+*** Do
+1. basic
+   - input
+     #+begin_src haskell
+       let foo xs =
+         a <- xs;
+         more-comp;
+         pure a
+     #+end_src
+   - output
+     #+begin_src lisp
+       (:defun foo (xs)
+         (:do (%<- a xs)
+              more-comp
+              (pure a)))
+     #+end_src
+*** Lambda
+1. basic
+   - input
+     #+begin_src haskell
+       let foo y =
+         \x -> x + y
+     #+end_src
+   - output
+     #+begin_src lisp
+       (:defun foo (y)
+         (:lambda (x) (:infix + x y)))
+     #+end_src
+*** Open
+1. basic
+   - input
+     #+begin_src haskell
+       let foo y =
+         open Prelude in
+         x + y
+     #+end_src
+   - output
+     #+begin_src lisp
+       (:defun foo (y)
+         (:open-in Prelude
+            (:infix + x y)))
+     #+end_src
+*** Parens
+1. basic
+   - input
+     #+begin_src haskell
+       let foo y = (y + 3) * 9
+     #+end_src
+   - output
+     #+begin_src lisp
+       (:defun foo (y)
+         (:infix * (:paren (:infix + y 3)) 9))
+     #+end_src
+*** Block
+1. basic
+   - input
+     #+begin_src haskell
+       let foo y = 3 * begin y + y end
+     #+end_src
+   - output
+     #+begin_src lisp
+       (:defun foo (y)
+         (:infix * 3 (:progn (:infix + y y))))
+     #+end_src
+*** Primitive
+1. basic
+   - input
+     #+begin_src haskell
+       let add = %Michelson.add
+     #+end_src
+   - output
+     #+begin_src lisp
+       (:defun add ()
+         (:primitive Michelson.add))
+     #+end_src
+*** Cond
+1. basic
+   - input
+     #+begin_src haskell
+       let add x =
+         if | x == 3 = 1
+            | x == 4 = 5
+            | else   = 0
+     #+end_src
+   - output
+     #+begin_src lisp
+       (:defun add (x)
+         (:cond ((:infix == x 3) 1)
+                ((:infix == x 4) 5)
+                (else            0)))
+     #+end_src
+* Notes
+** Design Principles
+Overall the goal of the S-expression incantation is twofold.
+
+1) An unambiguous syntax that is faithful to the original ML syntax
```

Comment:
What does it mean to be faithful? Is there a bijection (ideally)?

Reply:
Yes we preserve properties. Though when I wrote this i was using more colloquial meaning of the word

Reply:
Can you be more specific about what properties are preserved?